capsules/src/dals/README.md

```diff
@@ -0,0 +1,100 @@
+standard  communication protocol (BLE, Zigbee, USB, etc.) which provides "acks" or a similar service
+to lower data rates.  This data is then pushed 
+segment-by-segment through **DALS** which places it in an empty space in flash space and
+then sets up the process-specific elements for the newly-loaded app.
+
+## Design
+
+The design of **DALS** consists of a series of capsules with one capsule, the **AppLoader**, considered
+to be trusted.  This main trusted capsule is the heart of **DALS** and contains the code which collects
+the raw app data from another capsule, moves it through a pipeline of steps, places the
+resulting app binary in an empty slot in flash space, and, finally, sets up its process-memory region
+so the scheduler can run it at the appropriate time.
+
+Basic Pipeline at the core of **DALS**.
+
+
+Collect raw data from peripheral capsule (BLE, Zigbee, etc.) -----> Decompress Data  ----->  Find empty slot in flash --- 
+
+---> Verify Data ------> Validate Data ----> Set up process-specific structures (stack, heap, etc.)
+
+
+The trusted `AppLoader` capsule interfaces with 4 differennt peripheral capsules. These
+peripheral capsules consist of the **AppLoaderClient**, **Decompressor**, **Verifier**, and 
+**Validator** capsules.  Each capsule is named after the trait or interface it should implement.
+These interfaces are defined in `interfaces.rs`.
+
+The way these capsules are interconnected is diagrammed below.
+
+
+                                            AppLoaderClient
+                                                 |
+                                                 |
+                                                 |
+                                                 |
+                         Validator ---------  AppLoader  ----------  Decompressor
+                                                 |
+                                                 |
+                                                 |
+                                                 |
+                                              Verifier
+                                        
+
+There exists a circular reference between the **AppLoader** capsule and each of the peripheral
+capsules.  This enables the **AppLoader** to call functions implemented by a peripheral capsule and 
+vice versa.
+
+The peripheral modules are detailed below.  Please see `interfaces.rs` for more information on functions and such.
+
+
+## AppLoaderClient
+
+This capsule implements `pub trait AppLoaderClient<'a>`.  This capsule can be any capsule
+which receives data via standard communication protocols for embedded systems such as BLE, 
+Zigbee, USB.  This capsule's main function is to send this raw received data buffer-by-buffer
+to the **AppLoader** capsule.  In order to prevent bottleneck issues, this capsule must send an ack 
+only after its buffer is returned to it from the **AppLoader**.
```

Comment:
This may not be possible for arbitrary communication protocols. For example 802.15.4 expects an ACK within 1-2 ms or an automatic retry will be triggered, and Zigbee will give up on a MAC layer retransmissions (after sending 5x) after 37 ms. It seems that this realistically will create a requirement for higher layer flow control (i.e. use of TCP or similar).

Even for BLE, I don't think our current BLE stack supports L2CAP flow control, which is the only way I know of to achieve this.

Reply:
Sorry for missing this.  @bradjc and I seem converged on the idea of leaving this "traffic control" idea up to the actual protocol instead.  Obviously since there does not seem to be a one size fit all solution.  For protocols such as 802.15.4 for which timing is an issue, we think it is best to leave this problem up to the system designer.